plugin/src/compilation/decompiler.ml

```diff
@@ -140,85 +193,107 @@ let try_rel (trm : constr) : int option =
   | Rel i -> Some i
   | _ -> None
 
-(* Guard fails if the rel index occurs in the term. *)
-let noccur (i : int) (trm : constr) : unit option =
-  if Vars.noccurn i trm then Some () else None
+(* Monadic guard for option. *)
+let guard (b : bool) : unit option =
+  if b then Some () else None
 
-let rec first_pass env trm =
+(* Performs the bulk of decompilation on a proof term. 
+   Returns a list of tactics. *)
+let rec first_pass env sigma trm =
   (* Apply single beta reduction to terms that *might*
        be in eta expanded form. *)
   let trm = Reduction.whd_betaiota env trm in
-  let choose f x = Option.default [Apply (env, trm)] (f x env) in
+  let choose f x = Option.default [Apply (env, trm)] (f x (env, sigma)) in
   match kind trm with
   (* "fun x => ..." -> "intro x." *)
   | Lambda (n, t, b) ->
-     let in_env = get_pushed_names env in
-     let name = match n with
-       | Anonymous -> Id.of_string "H"
-       | Name n -> n in
-     let name = fresh_id_in_env in_env name env in
+     let name = fresh_name env n in
      let env = push_local (Name name, t) env in
-     Intro name :: first_pass env b
+     Intro name :: first_pass env sigma b
   (* Match on well-known functions used in the proof. *)
   | App (f, args) ->
-     choose (rewrite <|> left <|> right <|> split) (f, args)
+     choose (rewrite <|> induction <|> left <|> right <|> split) (f, args)
   (* Hypothesis transformations or generation tactics. *)
   | LetIn (n, valu, typ, body) ->   
      choose (rewrite_in <|> apply_in <|> pose) (n, valu, typ, body)
   (* Remainder of body, simply apply it. *)
   | _ -> [Apply (env, trm)]
 
-and rewrite (f, args) env : tact list option =
+(* Application of a equality eliminator. *)
+and rewrite (f, args) (env, sigma) : tact list option =
   try_rewrite f args >>= fun (left, prf, hyp) ->
-  Some (Rewrite (env, prf, left) :: first_pass env hyp)
+  Some (Rewrite (env, prf, left) :: first_pass env sigma hyp)
+
+(* Applying an eliminator for induction on a hypothesis in context. *)
+and induction (f, args) (env, sigma) : tact list option =
+  guard (is_elim env f) >>= fun _ ->
+  guard (not (is_rewrite f)) >>= fun _ ->
+  let app = mkApp (f, args) in
+  let sigma, ind = deconstruct_eliminator env sigma app in
+  let ind_args = ind.final_args in
+  (* Supports only 1 argument at the moment. *)
```

Comment:
Why support only one argument for now? I think you should be able to get away with inducting over the last argument in `ind_args`, since the other arguments to induction will be indices.

Reply:
Though, just like what you observed with `rewrite`, you will need to deal somehow with extra arguments at the end when the motive is dependent.